pkcs11/pkcs11.go

```diff
@@ -1687,3 +1734,160 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
 	}
 	return priv, nil
 }
+
+func (r *rsaPrivateKey) WithHash(hash crypto.Hash) (*rsaPrivateKey, error) {
+	r.hash = &hash
+	return r, nil
+}
+
+func (r *rsaPrivateKey) getHash() *crypto.Hash {
+	if r.hash != nil {
+		return r.hash
+	}
+	// Initialized to SHA256 if no hash specified
+	hash := crypto.SHA256
+	return &hash
+}
+
+func (r *rsaPrivateKey) WithPublicKeyHandle(o Object) *rsaPrivateKey {
+	r.pubH = &o
+	return r
+}
+
+func (r *rsaPrivateKey) encryptRSA(data []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam := (C.CK_RSA_PKCS_OAEP_PARAMS_PTR)(C.malloc(C.sizeof_CK_RSA_PKCS_OAEP_PARAMS))
+	defer C.free(unsafe.Pointer(cParam))
+
+	switch *hash {
+	case crypto.SHA256:
+		cParam.hashAlg = C.CKM_SHA256
+		cParam.mgf = C.CKG_MGF1_SHA256
+	case crypto.SHA384:
+		cParam.hashAlg = C.CKM_SHA384
+		cParam.mgf = C.CKG_MGF1_SHA384
+	case crypto.SHA512:
+		cParam.hashAlg = C.CKM_SHA512
+		cParam.mgf = C.CKG_MGF1_SHA512
+	case crypto.SHA1:
+		cParam.hashAlg = C.CKM_SHA_1
+		cParam.mgf = C.CKG_MGF1_SHA1
+	default:
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", hash)
+	}
```

Comment:
Let's make this a helper function

Reply:
Actually, build the entire cParam object in a helper.

Reply:
Done